nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java

```diff
@@ -0,0 +1,232 @@
+        if (stateMap.getVersion() == -1L || stateMap.get(CURRENT_TIMESTAMP) == null || stateMap.get(CURRENT_KEY_PREFIX) == null) {
+            currentTimestamp = 0L;
+            currentKeys = new HashSet<>();
+        } else {
+            currentTimestamp = Long.parseLong(stateMap.get(CURRENT_TIMESTAMP));
+            currentKeys = extractKeys(stateMap);
+        }
+    }
+
+    private void persistState(final ProcessContext context) {
+        Map<String, String> state = new HashMap<>();
+        state.put(CURRENT_TIMESTAMP, String.valueOf(currentTimestamp));
+        int i = 0;
+        for (String key : currentKeys) {
+            state.put(CURRENT_KEY_PREFIX+i, key);
+        }
+        try {
+            context.getStateManager().setState(state, Scope.CLUSTER);
+        } catch (IOException ioe) {
+            getLogger().error("Failed to save cluster-wide state. If NiFi is restarted, data duplication may occur", ioe);
+        }
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        try {
+            restoreState(context);
+        } catch (IOException ioe) {
+            getLogger().error("Failed to restore processor state; yielding", ioe);
+            context.yield();
+            return;
+        }
+
+        final long startNanos = System.nanoTime();
+        final String bucket = context.getProperty(BUCKET).evaluateAttributeExpressions().getValue();
+
+        final AmazonS3 client = getClient();
+        int listCount = 0;
+        long maxTimestamp = 0L;
+        String delimiter = context.getProperty(DELIMITER).getValue();
+        String prefix = context.getProperty(PREFIX).evaluateAttributeExpressions().getValue();
+
+        ListObjectsRequest listObjectsRequest = new ListObjectsRequest().withBucketName(bucket);
+        if (delimiter != null && !delimiter.isEmpty()) {
+            listObjectsRequest.setDelimiter(delimiter);
+        }
+        if (prefix != null && !prefix.isEmpty()) {
+            listObjectsRequest.setPrefix(prefix);
+        }
+
+        ObjectListing objectListing;
+        do {
+            objectListing = client.listObjects(listObjectsRequest);
+            for (S3ObjectSummary objectSummary : objectListing.getObjectSummaries()) {
+                long lastModified = objectSummary.getLastModified().getTime();
+                if (lastModified < currentTimestamp
+                        || lastModified == currentTimestamp && currentKeys.contains(objectSummary.getKey())) {
+                    continue;
+                }
+
+                // Create the attributes
+                final Map<String, String> attributes = new HashMap<>();
+                attributes.put(CoreAttributes.FILENAME.key(), objectSummary.getKey());
+                attributes.put("s3.bucket", objectSummary.getBucketName());
+                attributes.put("s3.owner", objectSummary.getOwner().getId());
```

Comment:
Seem to be getting NPEs fairly consistency regardless if the S3 bucket has files put into it from NiFi or those already existing in scenarios where a bucket may be publicly available.  Constraining permissions to an account using associated credentials seems to remedy this issue.
```
java.lang.NullPointerException: null
        at org.apache.nifi.processors.aws.s3.ListS3.onTrigger(ListS3.java:195) ~[nifi-aws-processors-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1139) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:139) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:1) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:124) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_77]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_77]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_77]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_77]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_77]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_77]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_77]
```